streams/shared/src/main/scala/zio/stream/ZStream.scala

```diff
@@ -2424,6 +2452,61 @@ object ZStream {
       }
   }
 
+  private[stream] sealed abstract class Structure[-R, +E, +A] {
+    def process: ZManaged[R, E, ZIO[R, Option[E], A]]
+  }
+
+  private[stream] object Structure {
+    final case class Iterator[-R, +E, +A](process: ZManaged[R, E, Pull[R, E, A]]) extends Structure[R, E, A]
+    final case class Concat[-R, +E, +A](hd: Structure[R, E, A], tl: () => Structure[R, E, A])
+        extends Structure[R, E, A] {
+      val process: ZManaged[R, E, Pull[R, E, A]] = {
+        def go(
+          doneRef: Ref[Boolean],
+          currPull: Ref[Pull[R, E, A]],
+          nextPull: Ref[Option[() => Structure[R, E, A]]],
+          switchPull: ZManaged[R, E, Pull[R, E, A]] => ZIO[R, E, Pull[R, E, A]]
+        ): ZIO[R, Option[E], A] =
+          doneRef.get.flatMap { done =>
+            if (done) IO.fail(None)
+            else
+              currPull.get.flatten.foldM(
```

Comment:
You can use `catchAll` to avoid having to deal with the happy path.

Reply:
Right!